spire-flex/tests/agentHealthCheckBindPort.yaml

```diff
@@ -0,0 +1,103 @@
+suite: test .agent.healthCheck.bindPort 9091
+set:
+  Release.Name: devcluster
+templates:
+  - agent-configmap.yaml
+  - agent-daemonset.yaml
+tests:
+  - it: "Should render the correct agent_config file"
+    template: agent-configmap.yaml
+    set:
+      agent.healthCheck.bindPort: 9091
+    asserts:
+      - containsDocument:
+          apiVersion: "v1"
+          kind: "ConfigMap"
+          name: "RELEASE-NAME-agent-config"
+          namespace: "NAMESPACE"
+      - equal:
```

Comment:
Concerned about checking the entire content of the configfile verbatim with every change to the codebase. The output will change and every test will need to be updated. This would quickly get unmanageable?

Reply:
Blame YAML.  There's no construct that is "part" of a multi-line string, and we don't want to rewrite helm unittest to do HCL subtesting within a string.

I don't think that the effort will be unmanagable, it will just be large.  Fortunately when the tests fail it will be easy to identify the differences, as that's the output of the unit test failure.  It also will let a person updating the unit tests know if they need to test for additional values and clue reviewers when one subsystem is stepping on another's toes.

If we can't accept this, we can't accept unit testing.  It comes hand-in-hand, and the benefits are "no surprises in our changes" which also reduces maintenance effort.  I'd rather sed / vi a few dozen files than find out the hard way we are introducing bugs.

Reply:
I just had to update all of the unit tests to adjust a container name.  It took all of about 40 seconds, even though it was 8 files, because it was just a vim macro.
If I had to adjust this with an IDE, it would have been a replace operation.  I think the "too much work" is the wrong way of thinking about things.  We have lots of tools (including sed!) to quickly replace items.

Reply:
False dichotomy. If we can't do it with helm-unittest, maybe helm-unittest is unsuitable, not that unit testing is unstable.

You had me research other unit testing frameworks and this one:
https://github.com/spiffe/helm-charts/blob/main/tests/unit/spire_test.go#L55
could easily parse the yaml/hcl/whatever from the configmap and then see if it has the properties expected.

Reply:
Also, 40 seconds isn't too bad I agree, but this is also for 1 feature. what happens when there are hundreds of configurables? Tweaking one setting in a pr could cause thousands of unit tests to all need updating. Not sure it will be as straight forward as a sed in future pr's.

Reply:
Seems like this conversation has drifted into items that can't be fixed by changing the code.

I'm not sure about the 40 seconds on your end.  In my end, my tests are running in less than a second, about 120 to 140 ms.  I wonder if your caching is broken, and you're downloading the tools each run.

Reply:
> I just had to update all of the unit tests to adjust a container name. It took all of about 40 seconds, even though it was 8 files, because it was just a vim macro

The 40 seconds was in response to this. Developer effort,  Not runtime length. As more tests of this type get added, the config file they will generate will differ test to test. The more of those we have, the more painful it will be when a new feature is added that changes what shows up in the config file by default. For example. We add the 'db' config to the configmap. as its always going to be in the config file, every test in the tree now needs updating to deal with the configmap change. I'm a bit concerned how much effort that may be for new contributions. each new contribution will require more and more work.